core/trino-main/src/main/java/io/trino/operator/join/JoinHashSupplier.java

```diff
@@ -79,9 +82,29 @@ public JoinHashSupplier(
             positionLinksFactoryBuilder = ArrayPositionLinks.builder(addresses.size());
         }
 
-        this.pages = channelsToPages(channels);
+        long additionalPagesRetainedSizeInBytes = 0;
+        // Currently PageIndex is retained by the HashBuilderOperator after JoinHash is created.
+        // JoinHash has to account for all data structures retained by PageIndex as after JoinHash is created memory retained by PageIndex is no longer accounted for.
+        additionalPagesRetainedSizeInBytes += sizeOf(pagePositionCount.elements());
+
+        ImmutableList.Builder<Page> pages = ImmutableList.builderWithExpectedSize(pagePositionCount.size());
+        for (int currentPage = 0; currentPage < pagePositionCount.size(); currentPage++) {
```

Comment:
Maybe instead inlining and modifying `channelsToPages` method you can use the method and postprocess obtained collection to calculate how much memory you need to account for on top of blocks. Sth like
```java
        this.pages = JoinUtils.channelsToPages(channels);
        for (Page page : pages) {
            additionalPagesRetainedSizeInBytes += page.getRetainedSizeInBytes();
            for (int channelId = 0; channelId < channels.size(); channelId++) {
                additionalPagesRetainedSizeInBytes -= page.getBlock(channelId).getRetainedSizeInBytes();
            }
        }
        this.additionalPagesRetainedSizeInBytes = additionalPagesRetainedSizeInBytes;
```
?